File: rust/src/nsenter.rs
//! Namespace enter helpers for rootful sandbox `popen()`.
//!
//! Replaces the `nsenter` subprocess with direct `setns()` + `chroot()`
//! syscalls, called from a Python `preexec_fn`.
//!
//! Note: PID namespace is **not** entered because `setns(CLONE_NEWPID)`
//! only affects future children, not the calling process or its exec.
//! This matches the userns `preexec_fn` behavior.

use std::ffi::CString;
use std::io;

/// Enter the mount namespace of `target_pid`, chroot to its root,
/// and chdir to "/".
///
/// Intended to be called from a `preexec_fn` (after fork, before exec).
/// The calling process must have `CAP_SYS_ADMIN` (i.e. running as root).
///
/// Like real `nsenter`, we open `/proc/{pid}/root` as an fd **before**
/// entering the mount namespace, then `fchdir` + `chroot(".")` after.
/// This avoids the problem of /proc becoming inaccessible after setns.
pub fn nsenter_preexec(target_pid: i32) -> io::Result<()> {
    // 1. Open the target's root fd BEFORE entering mount namespace.
    //    After setns, the host's /proc may not be visible.
    let root_path = format!("/proc/{target_pid}/root");
    let c_root = CString::new(root_path.as_bytes())
        .map_err(|e| io::Error::new(io::ErrorKind::InvalidInput, e))?;
    let root_fd = unsafe { libc::open(c_root.as_ptr(), libc::O_RDONLY | libc::O_CLOEXEC) };
    if root_fd < 0 {
        return Err(io::Error::last_os_error());
    }

    // 2. Open mount namespace fd.
    let mnt_ns = format!("/proc/{target_pid}/ns/mnt");
    let c_mnt = CString::new(mnt_ns.as_bytes())
        .map_err(|e| io::Error::new(io::ErrorKind::InvalidInput, e))?;
    let mnt_fd = unsafe { libc::open(c_mnt.as_ptr(), libc::O_RDONLY | libc::O_CLOEXEC) };
    if mnt_fd < 0 {
        unsafe { libc::close(root_fd) };
        return Err(io::Error::last_os_error());
    }

    // 3. Enter mount namespace.
    let rc = unsafe { libc::setns(mnt_fd, libc::CLONE_NEWNS) };
    unsafe { libc::close(mnt_fd) };
    if rc < 0 {
        unsafe { libc::close(root_fd) };
        return Err(io::Error::last_os_error());
    }

    // 4. fchdir to the pre-opened root fd, then chroot(".").
    if unsafe { libc::fchdir(root_fd) } < 0 {
        unsafe { libc::close(root_fd) };
        return Err(io::Error::last_os_error());
    }
    unsafe { libc::close(root_fd) };

    let c_dot = CString::new(".").unwrap();
    if unsafe { libc::chroot(c_dot.as_ptr()) } < 0 {
        return Err(io::Error::last_os_error());
    }

    // 5. chdir to /.
    let c_slash = CString::new("/").unwrap();
    if unsafe { libc::chdir(c_slash.as_ptr()) } < 0 {
        return Err(io::Error::last_os_error());
    }

    Ok(())
}

/// Enter user + mount namespace of `target_pid`, chroot to `rootfs`,
/// and chdir to `working_dir`.
///
/// Used by userns `popen()`.  The difference from [`nsenter_preexec`] is
/// that we also join the user namespace (needed for rootless sandboxes)
/// and chroot to an explicit rootfs path instead of /proc/{pid}/root.
///
/// For userns sandboxes, the rootfs path is on the host filesystem
/// (not inside the target's mount namespace), so we open namespace fds
/// first, then enter both namespaces, then chroot to the host-side path.
pub fn userns_preexec(target_pid: i32, rootfs: &str, working_dir: &str) -> io::Result<()> {
    // 1. Open namespace fds BEFORE entering any namespace.
    let user_ns = format!("/proc/{target_pid}/ns/user");
    let c_user = CString::new(user_ns.as_bytes())
        .map_err(|e| io::Error::new(io::ErrorKind::InvalidInput, e))?;
    let user_fd = unsafe { libc::open(c_user.as_ptr(), libc::O_RDONLY | libc::O_CLOEXEC) };
    if user_fd < 0 {
        return Err(io::Error::last_os_error());
    }

    let mnt_ns = format!("/proc/{target_pid}/ns/mnt");
    let c_mnt = CString::new(mnt_ns.as_bytes())
        .map_err(|e| io::Error::new(io::ErrorKind::InvalidInput, e))?;
    let mnt_fd = unsafe { libc::open(c_mnt.as_ptr(), libc::O_RDONLY | libc::O_CLOEXEC) };
    if mnt_fd < 0 {
        unsafe { libc::close(user_fd) };
        return Err(io::Error::last_os_error());
    }

    // 2. Enter user namespace.
    let rc = unsafe { libc::setns(user_fd, libc::CLONE_NEWUSER) };
    unsafe { libc::close(user_fd) };
    if rc < 0 {
        unsafe { libc::close(mnt_fd) };
        return Err(io::Error::last_os_error());
    }

    // 3. Enter mount namespace.
    let rc = unsafe { libc::setns(mnt_fd, libc::CLONE_NEWNS) };
    unsafe { libc::close(mnt_fd) };
    if rc < 0 {
        return Err(io::Error::last_os_error());
    }

    // 4. chroot to the rootfs.
    let c_root =
        CString::new(rootfs).map_err(|e| io::Error::new(io::ErrorKind::InvalidInput, e))?;
    if unsafe { libc::chroot(c_root.as_ptr()) } < 0 {
        return Err(io::Error::last_os_error());
    }

    // 5. chdir to working_dir.
    let c_wd =
        CString::new(working_dir).map_err(|e| io::Error::new(io::ErrorKind::InvalidInput, e))?;
    if unsafe { libc::chdir(c_wd.as_ptr()) } < 0 {
        return Err(io::Error::last_os_error());
    }

    Ok(())
}

File: go/internal/nsenter/nsenter.go
// Package nsenter provides namespace enter helpers for sandbox popen().
package nsenter

import (
	"fmt"
	"os"
	"syscall"

	"golang.org/x/sys/unix"
)

// NsenterPreexec enters the mount namespace of targetPid, chroots to its root,
// and chdirs to "/". Used for rootful popen().
func NsenterPreexec(targetPid int) error {
	// 1. Open root fd BEFORE entering mount namespace
	rootPath := fmt.Sprintf("/proc/%d/root", targetPid)
	rootFd, err := unix.Open(rootPath, unix.O_RDONLY|unix.O_CLOEXEC, 0)
	if err != nil {
		return err
	}

	// 2. Open mount namespace fd
	mntNs := fmt.Sprintf("/proc/%d/ns/mnt", targetPid)
	mntFd, err := unix.Open(mntNs, unix.O_RDONLY|unix.O_CLOEXEC, 0)
	if err != nil {
		unix.Close(rootFd)
		return err
	}

	// 3. Enter mount namespace
	if err := unix.Setns(mntFd, unix.CLONE_NEWNS); err != nil {
		unix.Close(mntFd)
		unix.Close(rootFd)
		return err
	}
	unix.Close(mntFd)

	// 4. fchdir + chroot
	if err := unix.Fchdir(rootFd); err != nil {
		unix.Close(rootFd)
		return err
	}
	unix.Close(rootFd)

	if err := unix.Chroot("."); err != nil {
		return err
	}

	// 5. chdir to /
	return unix.Chdir("/")
}

// UsersPreexec enters user+mount namespace, chroots to rootfs, chdirs to workingDir.
func UsersPreexec(targetPid int, rootfs, workingDir string) error {
	// 1. Open namespace fds
	userNs := fmt.Sprintf("/proc/%d/ns/user", targetPid)
	userFd, err := unix.Open(userNs, unix.O_RDONLY|unix.O_CLOEXEC, 0)
	if err != nil {
		return err
	}

	mntNs := fmt.Sprintf("/proc/%d/ns/mnt", targetPid)
	mntFd, err := unix.Open(mntNs, unix.O_RDONLY|unix.O_CLOEXEC, 0)
	if err != nil {
		unix.Close(userFd)
		return err
	}

	// 2. Enter user namespace
	if err := unix.Setns(userFd, unix.CLONE_NEWUSER); err != nil {
		unix.Close(userFd)
		unix.Close(mntFd)
		return err
	}
	unix.Close(userFd)

	// 3. Enter mount namespace
	if err := unix.Setns(mntFd, unix.CLONE_NEWNS); err != nil {
		unix.Close(mntFd)
		return err
	}
	unix.Close(mntFd)

	// 4. chroot
	if err := unix.Chroot(rootfs); err != nil {
		return err
	}

	// 5. chdir
	return unix.Chdir(workingDir)
}

// NsenterExec enters a namespace and executes a command.
// This replaces the Python preexec_fn pattern — Python calls this as a subprocess.
func NsenterExec(targetPid int, rootful bool, rootfs, workingDir string, cmdArgs []string) error {
	if rootful {
		if err := NsenterPreexec(targetPid); err != nil {
			return err
		}
	} else {
		if err := UsersPreexec(targetPid, rootfs, workingDir); err != nil {
			return err
		}
	}

	// exec the command
	bin, err := LookPath(cmdArgs[0])
	if err != nil {
		return err
	}
	return syscall.Exec(bin, cmdArgs, os.Environ())
}

// LookPath finds a binary in PATH inside the chroot.
func LookPath(name string) (string, error) {
	if name[0] == '/' {
		return name, nil
	}
	paths := []string{"/usr/local/sbin", "/usr/local/bin", "/usr/sbin", "/usr/bin", "/sbin", "/bin"}
	pathEnv := os.Getenv("PATH")
	if pathEnv != "" {
		paths = append([]string{}, splitPath(pathEnv)...)
	}
	for _, dir := range paths {
		full := dir + "/" + name
		if _, err := os.Stat(full); err == nil {
			return full, nil
		}
	}
	return "", fmt.Errorf("executable not found: %s", name)
}

func splitPath(path string) []string {
	var result []string
	for _, p := range splitString(path, ':') {
		if p != "" {
			result = append(result, p)
		}
	}
	return result
}

func splitString(s string, sep byte) []string {
	var result []string
	for {
		i := indexByte(s, sep)
		if i < 0 {
			result = append(result, s)
			break
		}
		result = append(result, s[:i])
		s = s[i+1:]
	}
	return result
}

func indexByte(s string, b byte) int {
	for i := 0; i < len(s); i++ {
		if s[i] == b {
			return i
		}
	}
	return -1
}
